File: fact-app/consumer.py
"""
Kafka consumer: on tenant.signups, if 'fact-app' is in services, create tenant DB and register in TenantDB.
"""
import json
import logging
from confluent_kafka import Consumer, KafkaError
from config import Config
from tenant_pool import create_tenant_database

logger = logging.getLogger(__name__)

def run_consumer():
    c = Consumer({
        "bootstrap.servers": Config.KAFKA_BOOTSTRAP,
        "group.id": Config.APP_NAME,
        "auto.offset.reset": "earliest",
    })
    c.subscribe([Config.KAFKA_TOPIC])
    try:
        while True:
            msg = c.poll(1.0)
            if msg is None:
                continue
            if msg.error():
                if msg.error().code() == KafkaError._PARTITION_EOF:
                    continue
                logger.error("Consumer error: %s", msg.error())
                continue
            try:
                evt = json.loads(msg.value().decode())
                tenant_id = evt.get("tenant_id")
                services = evt.get("services") or []
                if tenant_id and Config.APP_NAME in services:
                    create_tenant_database(tenant_id)
                    logger.info("Created tenant DB for %s", tenant_id)
            except Exception as e:
                logger.exception("Processing event: %s", e)
    finally:
        c.close()

File: org-app/internal/consumer/consumer.go
package consumer

import (
	"context"
	"encoding/json"
	"log"

	"github.com/IBM/sarama"
	"github.com/hybrid-tenancy-demo/org-app/internal/config"
	"github.com/hybrid-tenancy-demo/org-app/internal/db"
)

type TenantSignupEvent struct {
	TenantID string   `json:"tenant_id"`
	Name     string   `json:"name"`
	Services []string `json:"services"`
}

func Run(ctx context.Context, registry *db.Registry, cfg *config.Config) {
	c, err := sarama.NewConsumerGroup(cfg.KafkaBrokers, config.AppName, sarama.NewConfig())
	if err != nil {
		log.Fatalf("kafka consumer: %v", err)
	}
	defer c.Close()

	handler := &signupHandler{registry: registry, cfg: cfg}
	for {
		select {
		case <-ctx.Done():
			return
		default:
			if err := c.Consume(ctx, []string{cfg.KafkaTopic}, handler); err != nil {
				log.Printf("consume: %v", err)
			}
		}
	}
}

type signupHandler struct {
	registry *db.Registry
	cfg      *config.Config
}

func (h *signupHandler) Setup(sarama.ConsumerGroupSession) error   { return nil }
func (h *signupHandler) Cleanup(sarama.ConsumerGroupSession) error { return nil }

func (h *signupHandler) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
	for msg := range claim.Messages() {
		var evt TenantSignupEvent
		if err := json.Unmarshal(msg.Value, &evt); err != nil {
			continue
		}
		for _, svc := range evt.Services {
			if svc == config.AppName {
				if err := db.CreateTenantDatabase(h.cfg.DBHost, h.cfg.DBPort, h.cfg.DBUser, h.cfg.DBPassword, evt.TenantID); err != nil {
					log.Printf("failed to create tenant db %s: %v", evt.TenantID, err)
					return err
				}

				log.Printf("created tenant db %s", evt.TenantID)
				if err := h.registry.Register(evt.TenantID, h.cfg.DBHost); err != nil {
					log.Printf("failed to register tenant %s: %v", evt.TenantID, err)
					return err
				} 

				log.Printf("registered tenant %s", evt.TenantID)
				log.Printf("created org-app tenant DB for %s", evt.TenantID)
				break
			}
		}
		session.MarkMessage(msg, "")
	}
	return nil
}
